ENH/MNT: Add a common initial dataset creation routine

greglucas opened the pull request:
# Change Summary

## Overview

This adds a common interface for returning xarray datasets from a packet file. There is one dataset per apid, sorted by time, with derived values automatically converted.

Each dataset is essentially a L0.5, but my hope is that this will be easier for everyone to work with (especially for housekeeping dataset creation). Rather than everyone implementing something fairly similar but slightly different inside their packet iteration routine:
```python
for packet in packets:
    # do a bunch of checks like sorting, apid checks, extraction
```

It would be:

```python
datasets_by_apid = packet_file_to_datasets(packet_file, instrument_definition)
ds = datasets_by_apid[my_instrument.SCIENCE_APID]

... Then later do some work on the dataset
# The dataset is already referenced to our required epoch and is the length of the number of packets in the file with this apid
# This will allow easier iteration as the entire dataset is referenced to the proper dimension of the packet epoch.
ds["my_derirved_variable"] = do_something(ds["my_packet_variable"])
```

I updated SWAPI to use this new function, which simplified things IMO.

greglucas committed:
ENH/MNT: Add a common initial dataset creation routine

This adds a common interface for returning xarray datasets from
a packet file. There is one dataset per apid, sorted by time, with
derived values automatically converted.

tmplummer commented on imap_processing/utils.py:
```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
+                    )
+                    # TODO: Additional info from XTCE like units
+
+            # TODO: Include the header too? (packet.header | packet.data)
```
Generally, l1a is close to the unprocessed packet, so I would argue that similar to raw vs. derived value, the header should be included for l1a an not included for l1b.

greglucas replied:
OK, I'm happy to do that as well, then we could strip it later.

For the raw vs `derived_value`. Do we not want to use the XTCE functionality of `derived_value` and rather just do that ourselves in the code later? i.e. use the raw_to_eu conversion function instead.

tmplummer replied:
> For the raw vs `derived_value`. Do we not want to use the XTCE functionality of `derived_value` and rather just do that ourselves in the code later? i.e. use the raw_to_eu conversion function instead.

Good question. I recall a conversation about using the command and telemetry spreadsheet analog conversions to populate the XTCE so that we could essentially skip l1a for housekeeping. Maybe that decision has not been made and just sticking to the raw value is the appropriate thing to do for now?

greglucas replied:
I added an optional argument to it that defaults to `derived_value`, but allows one to get the raw value if we want too. I can remove it or change the default if you think this would be better as `raw` by default. I actually think we should use derived if we can, but I'm not wedded to that.

tmplummer commented on imap_processing/utils.py:
```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
+            coords={"epoch": time_data},
+        )
+        for key, value in attribute_dict[apid].items():
+            ds[key.lower()].attrs.update(value)
```
Is the assumption here that all of the required CDF attributes will be defined in the xtce? or would there be another round of updating the variable attributes in the dataset?

greglucas replied:
No, I view this as very much a "just give me an xarray dataset". Then I'll start appending metadata, adding more stuff to it etc... We could leave this out if it'd be easier to understand and use later, but I figured it might be useful for housekeeping datasets.

tmplummer replied:
Ok, great. I wonder if it is even useful to include the attributes defined in XTCE or if we should just centralize on using the YAML for all attributes.

greglucas replied:
I ended up removing the attributes and short/long descriptions. It simplified things, so I think makes sense for now.

tmplummer commented on imap_processing/utils.py:
```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
+    packet_parser = parser.PacketParser(packet_definition)
+
+    with open(packet_file, "rb") as binary_data:
+        packet_generator = packet_parser.generator(binary_data)
```
I like that this directly uses the generator rather than converting to a list.

tmplummer commented on imap_processing/utils.py:
```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
     return dataset
+
+
+def packet_file_to_datasets(
```
I see quite a bit of overlap with the `create_dataset` function. Some differences are:

- This function creates a dataset for each apid in the input file
- `create_dataset` has an option for including the packet header or not
- `create_dataset` brings in variable attributes from a source external to the xtce. Right now that source has not been updated to use the ImapCdfAttributes class.
- `create_dataset` allows for the option to skip certain fields in the packet.

greglucas replied:
I agree, this is heavily based on that function. Would you want all of these capabilities? My thought is that sometimes too many options is not great either and sometimes standardizing on a fixed layout is useful.

I'd actually rather not "skip" the specific fields and complicate this function, but rather just have a user do `dataset.drop("variable")` to remove whatever they want afterwards. So I kind of view this as wanting to give the user everything and then the user can start winnowing down to what they need or transforming from there.

If it would be useful to add some ignore keywords I am open to it though.

tmplummer replied:
I was really just making an observation about the overlap. I am in agreement that keeping this function simple makes sense.

tech3371 replied:
I agree. I think it's gonna save lot of repeated steps that we all are doing separately. This function seems much better. It's decomming, creating dataset and then returning it per apid. Then every processing that happens after that can decide what to keep and what to drop in terms of the actual data. And it's already in the format we need to write to cdf except we need to do minor things to add attrs stuff.

tmplummer commented on imap_processing/utils.py:
```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
     return dataset
+
+
+def packet_file_to_datasets(
```
It may be useful to optionally skip an apid. The motivation for this is packets with variable length binary data fields.

greglucas replied:
This handles variable length binary data!

I can see the argument here as well, but again I think that it might be easier for a user to just not access that apid in the dictionary lookup?

tmplummer reviewed (commented):
Overall, I like this implementation but I think there may need to be some features added in order to be able to replace all usages of the `create_dataset` function. I am also curious to see what it looks like to update variable attributes using the ImapCdfAttributes class.

greglucas committed:
MNT: Add option to not use derived value from XTCE

- Remove some attributes from XTCE to avoid duplication with yaml
- Add use_derived_value as a boolean option for whether or not to
  use the derived_value or keep it as the raw bits.

greglucas committed:
MNT: Change IMAP-Hi over to use new dataset accessor

This updates the Hi utilities to use the dataset creator. There
is still some work that can be done to improve and numpy broadcast
some of the routines, but this at least shows the basics.

greglucas commented:
@subagonsouth thanks for the comments. I updated `IMAP-Hi` now with the new function to remove the convert_time function as well. I hope that gives you an idea of how this will look for others, but I'm hoping others can make use of this and then hopefully we can even broadcast numpy array results in your Hi functions ;)

tech3371 commented on imap_processing/cdf/utils.py:
```diff
@@ -149,7 +151,7 @@ def write_cdf(dataset: xr.Dataset) -> xr.Dataset:
         repointing=repointing,
     )
-    file_path = science_file.construct_path()
+    file_path: Path = science_file.construct_path()
```
Was this done for a reason? Just curious

greglucas replied:
Because mypy was complaining later when it was used. I think our `imap-data-access` doesn't have type hints yet so this adds it for this one use.

tech3371 commented on imap_processing/utils.py:
```diff
@@ -228,29 +228,77 @@ def create_dataset(
+                # This is the first packet for this APID
+                data_dict[apid] = collections.defaultdict(list)
+
+            # TODO: Do we want to give an option to remove the header content?
+            packet_content = packet.data | packet.header
```
I do think it will be helpful to add that option in this function. Removing packet header content will be same across all packet and makes sense to do that here. As for removing data following header, it makes sense to do that in the individual processing code as you mentioned.

greglucas replied:
I am starting to be more of the mindset that we don't want to be adding options just to have them, is there a specific instrument/use-case you know of where we want both cases. So for now, I think keeping all packet content seems reasonable and then downstream people can remove the header if they want. But it seems like more people wanted the header than not?

tech3371 replied:
I have not heard strong preference of wanting to remove packet header. I think it came up in past discussion and brought it up just in case. That makes sense and works for me.

tech3371 commented on imap_processing/utils.py:
```diff
@@ -228,29 +228,77 @@ def create_dataset(
+    dataset_by_apid = {}
+    # Convert each apid's data to an xarray dataset
+    for apid, data in data_dict.items():
+        # The time key is always the first key in the data dictionary on IMAP
+        time_key = next(iter(data.keys()))
```
I am little confused about the comment. Do you mean that the time key is always the first key if we don't include the packet header and only include the data itself? That is true if that's the case. But if we include packet header than it won't the first key, right? Just want to make sure I understood your comment

greglucas replied:
I agree it is different depending on the header, so I avoided that situation here by making it the first key right above this by appending the packet header after the data. So regardless if we want to use the header or not, the SHCOARSE will be the first entry in this content dict.
```python
packet_content = packet.data | packet.header
```

tech3371 replied:
I see. Nice! I totally missed the order in which you added those. That makes sense now. Thank you!

tech3371 commented on imap_processing/swapi/l1/swapi_l1.py:
```diff
@@ -475,7 +468,8 @@ def process_swapi_science(sci_dataset: xr.Dataset, data_version: str) -> xr.Data
     # epoch time. Should be same dimension as number of good sweeps
     epoch_time = good_sweep_sci["epoch"].data.reshape(total_full_sweeps, 12)[:, 0]
-    epoch_converted_time = met_to_j2000ns(epoch_time)
+    # epoch_converted_time = met_to_j2000ns(epoch_time)
```
remove?

greglucas replied:
Done.

tech3371 commented on imap_processing/swapi/l1/swapi_l1.py:
```diff
@@ -532,8 +526,9 @@ def process_swapi_science(sci_dataset: xr.Dataset, data_version: str) -> xr.Data
     )
 
     # L1 quality flags
+    # TODO: Should these be kept in raw format rather than derived into strings?
```
keep it raw for now. I will ask SWAPI team if we want in derived value.

greglucas replied:
OK, I reverted this change and set `use_derived_value=False` for SWAPI. See the latest commit for the update.

tech3371 replied:
thank you!

tech3371 commented on imap_processing/tests/swapi/test_swapi_l1.py:
```diff
@@ -208,10 +203,7 @@ def test_swapi_algorithm(decom_test_data):
-    science_data = grouped_data[SWAPIAPID.SWP_SCI]
-    sorted_packets = sort_by_time(science_data, "SHCOARSE")
-    ds_data = create_dataset(sorted_packets, include_header=False)
+    ds_data = decom_test_data[SWAPIAPID.SWP_SCI]
```
nice!

tech3371 commented on imap_processing/tests/swapi/test_swapi_l1.py:
```diff
@@ -333,5 +325,6 @@ def test_swapi_l1_cdf():
     # hk cdf file
     cdf_filename = "imap_swapi_l1_hk_20100101_v001.cdf"
-    cdf_path = write_cdf(processed_data[1])
+    # Ignore ISTP checks for HK data
+    cdf_path = write_cdf(processed_data[1], istp=False)
```
nice!

greglucas committed:
MNT: Remove unused SWAPI epoch variable

greglucas committed:
MNT: Revert to using raw values for SWAPI

This removes the change to SWAPI to use the derived values. It also
abstracts the Enums out of the tests so future updates if we want
to use the derived values will be easier.

tech3371 reviewed (approved):
Thank you for doing this! Looks great!

greglucas commented:
@subagonsouth, as an FYI, I did update some of Hi's code in this as well and left some TODO comments here. Let me know if you want me to remove/change any of the Hi updates.

tmplummer commented on imap_processing/hi/l1a/histogram.py:
```diff
@@ -35,30 +33,34 @@
 
+    # unpack the packets data into the Dataset
+    # (npackets, 24 * 90 * 12)
+    # TODO: Look into avoiding the for-loops below
```
Yes... Good recommendation!

tmplummer commented on imap_processing/hi/l1a/histogram.py:
```diff
@@ -35,30 +33,34 @@
-        dataset.epoch.data[i_epoch] = met_to_j2000ns(packet.data["CCSDS_MET"].raw_value)
-        dataset.ccsds_met[i_epoch] = packet.data["CCSDS_MET"].raw_value
-        dataset.esa_stepping_num[i_epoch] = packet.data["ESA_STEP"].raw_value
+    # TODO: Move into the allocate dataset function?
```
These changes look fine for now. The `allocate_histogram_dataset` function was really just my way of avoiding having an intermediate data storage. I wanted to go directly from the packet into the xr.DataSet arrays. With your new function, this is no longer achieving that goal. I will write a ticket to look at this Hi code and address the TODOs.
